packages/astro/src/transitions/router.ts

```diff
@@ -637,8 +637,28 @@ if (inBrowser) {
 		originalLocation = new URL(location.href);
 		addEventListener('popstate', onPopState);
 		addEventListener('load', onPageLoad);
-		if ('onscrollend' in window) addEventListener('scrollend', onScroll);
-		else addEventListener('scroll', throttle(onScroll, 350), { passive: true });
+		if ('onscrollend' in window) addEventListener('scrollend', onScrollEnd);
+		// HACK: emulate "onscrollend" behavior by checking scroll positions with
+		// `requestAnimationFrame` every 350ms.
+		else {
+			let timestamp: number, y = scrollY, x = scrollX;
+			const cb = (now: number) => {
+				timestamp ??= now;
+				if (
+					now - timestamp > 350
+					&& y === scrollY && x === scrollX
+					&& (y !== history.state.scrollY || x !== history.state.scrollX)
```

Comment:
if a popstate event took place in the meanwhile, `history.state` has changed to that of the new page